src/utils/daos_hdlr.c

```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+				       "%s", ptr);
+			if (len > key_buf_len) {
+				fprintf(stderr, "dkey buffer is too small\n");
+				D_FREE(dkey);
+				D_GOTO(out, rc = 1);
+			}
+			d_iov_set(&diov, (void *)dkey, dkey_kds[j].kd_key_len);
+
+			/* enumerate and parse akeys */
+			rc = cont_clone_list_akeys(src_oh, dst_oh, diov);
+			if (rc != 0) {
+				fprintf(stderr, "failed to list akeys: %d\n",
+					rc);
+				D_FREE(dkey);
+				D_GOTO(out, rc);
+			}
+			ptr += dkey_kds[j].kd_key_len;
+			D_FREE(dkey);
+		}
+	}
+out:
+	D_FREE(small_key);
+	D_FREE(large_key);
+	return rc;
+}
+
+int
+cont_clone_hdlr(struct cmd_args_s *ap)
+{
+	int			rc = 0;
+	int			i = 0;
+	daos_cont_info_t	src_cont_info;
+	daos_cont_info_t	dst_cont_info;
+	daos_obj_id_t		oids[OID_ARR_SIZE];
+	daos_anchor_t		anchor;
+	uint32_t		oids_nr;
+	daos_handle_t		toh;
+	daos_epoch_t		epoch;
+	struct			dm_args ca = {0};
+	bool			is_posix_copy = false;
+	daos_handle_t		oh;
+	daos_handle_t		dst_oh;
+	struct file_dfs		src_cp_type = {0};
+	struct file_dfs		dst_cp_type = {0};
+	char			*src_str = NULL;
+	char			*dst_str = NULL;
+	size_t			src_str_len = 0;
+	size_t			dst_str_len = 0;
+	daos_epoch_range_t	epr;
+	bool			daos_no_prefix = true;
+	int			uuid_len = 128;
+
+	set_dm_args_default(&ca);
+	file_set_defaults_dfs(&src_cp_type);
+	file_set_defaults_dfs(&dst_cp_type);
+
+	src_str_len = strlen(ap->src);
+	D_STRNDUP(src_str, ap->src, src_str_len);
+	if (src_str == NULL) {
+		fprintf(stderr, "Unable to allocate memory for source path.");
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+	rc = dm_parse_path(&src_cp_type, src_str, src_str_len,
+			   &ca.src_p_uuid, &ca.src_c_uuid, daos_no_prefix);
+	if (rc != 0) {
+		fprintf(stderr, "failed to parse source path: %d\n", rc);
+		D_GOTO(out, rc = daos_errno2der(rc));
+	}
+
+	dst_str_len = strlen(ap->dst);
+	D_STRNDUP(dst_str, ap->dst, dst_str_len);
+	if (dst_str == NULL) {
+		fprintf(stderr, "Unable to allocate memory for dest path.");
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+	rc = dm_parse_path(&dst_cp_type, dst_str, dst_str_len,
+			   &ca.dst_p_uuid, &ca.dst_c_uuid, daos_no_prefix);
+	/* parse pool uuid if this fails, since it is not in form that
+	 * can be parsed by duns_resolve_path (i.e. dst=/$pool)
+	 */
+	if (rc != 0) {
+		uuid_generate(ca.dst_c_uuid);
+		ap->dst += 1;
+		D_STRNDUP(ca.dst, ap->dst, uuid_len);
+		if (ca.dst == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+		uuid_parse(ca.dst, ca.dst_p_uuid);
+		ap->dst -= 1;
+		/* not considered an error yet since this could
+		 * be a UNS path, dm_connect will check this
+		 * when it attempts to do a uns_path_create
+		 */
+		rc = 0;
+	} else {
+		rc = daos_pool_connect(ca.dst_p_uuid, ap->sysname,
+				       DAOS_PC_RW, &ca.dst_poh, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to destination "
+				"pool: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+		/* make sure this destination container doesn't exist already,
+		 * if it does, exit
+		 */
+		rc = daos_cont_open(ca.dst_poh, ca.dst_c_uuid, DAOS_COO_RW,
```

Comment:
@dsikich Doesn't dm_connect below connect to the pool and cont?

Reply:
@daltonbohning yes, but I figured there is no reason to enter that function and connect to the src if the destination already exists, since that is an error. Plus, fs copy uses dm_connect, and this is not an error for fs copy.

Reply:
@dsikich What I mean is that if we connect to the dst pool here, aren't we connecting to the dst pool again in dm_connect? When we connect the second time, the first handle is still open, right?

Reply:
@daltonbohning hmm well I suppose in dm_connect I only need to try and connect to the pool again if the dst uuid is null, but I think I do this because if this is a filesystem copy then the pool would not already be open. I can close the pool before entering dm_connect though, since doing it the other way will not work for fs copy.

Reply:
@dsikich Maybe `daos_handle_is_valid` could be useful in dm_connect to only connect if the handle is NOT valid. I.e. we haven't yet connected

Reply:
@daltonbohning it looks like that should work, I can add a check for a valid handle before calling connect on the destination